File: src/main/java/com/mohe/spring/entity/Place.java
package com.mohe.spring.entity;

import com.vladmihalcea.hibernate.type.array.ListArrayType;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.Type;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "places")
@Getter
@Setter
public class Place {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    @Column(precision = 10, scale = 8)
    private BigDecimal latitude;

    @Column(precision = 11, scale = 8)
    private BigDecimal longitude;

    @Column(name = "road_address", columnDefinition = "TEXT")
    private String roadAddress;

    @Column(name = "website_url", columnDefinition = "TEXT")
    private String websiteUrl;

    private BigDecimal rating;

    private Integer reviewCount;

    @Type(ListArrayType.class)
    @Column(name = "category", columnDefinition = "varchar[]")
    private List<String> category;

    @Type(ListArrayType.class)
    @Column(name = "keyword", columnDefinition = "varchar[]")
    private List<String> keyword;

    private Boolean parkingAvailable;

    private Boolean petFriendly;

    @Enumerated(EnumType.STRING)
    @Column(name = "crawl_status")
    private CrawlStatus crawlStatus = CrawlStatus.PENDING;

    @Enumerated(EnumType.STRING)
    @Column(name = "embed_status")
    private EmbedStatus embedStatus = EmbedStatus.PENDING;

    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    @OneToMany(mappedBy = "place", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PlaceDescription> descriptions = new ArrayList<>();

    @OneToMany(mappedBy = "place", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PlaceImage> images = new ArrayList<>();

    @OneToMany(mappedBy = "place", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PlaceBusinessHour> businessHours = new ArrayList<>();

    @OneToMany(mappedBy = "place", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PlaceSns> sns = new ArrayList<>();

    @OneToMany(mappedBy = "place", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PlaceReview> reviews = new ArrayList<>();

    @OneToMany(mappedBy = "place", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PlaceMenu> menus = new ArrayList<>();

    /**
     * 이미지 URL 리스트 (배치 처리용 임시 필드)
     * DB에 저장되지 않으며, Processor에서 Writer로 데이터 전달 시 사용
     */
    @Transient
    private List<String> imageUrls;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: src/main/kotlin/com/mohe/spring/entity/Place.kt
package com.mohe.spring.entity

import jakarta.persistence.*
import java.math.BigDecimal
import java.time.OffsetDateTime

@Entity
@Table(name = "places")
data class Place(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,
    
    @Column(nullable = false)
    val name: String,
    
    @Column(nullable = false)
    val title: String,
    
    val address: String? = null,
    val location: String? = null,
    
    @Column(precision = 10, scale = 8)
    val latitude: BigDecimal? = null,
    
    @Column(precision = 11, scale = 8)
    val longitude: BigDecimal? = null,
    
    @Column(precision = 10, scale = 2)
    val altitude: BigDecimal? = null,
    
    val category: String? = null,
    val description: String? = null,
    
    @Column(name = "image_url")
    val imageUrl: String? = null,
    
    @Convert(converter = StringArrayConverter::class)
    val images: List<String> = emptyList(),
    
    @Convert(converter = StringArrayConverter::class)
    val gallery: List<String> = emptyList(),
    
    @Column(name = "additional_image_count")
    val additionalImageCount: Int = 0,
    
    @Column(precision = 3, scale = 2)
    val rating: BigDecimal = BigDecimal.ZERO,
    
    @Column(name = "review_count")
    val reviewCount: Int = 0,
    
    @Column(name = "operating_hours")
    val operatingHours: String? = null,
    
    @Convert(converter = StringArrayConverter::class)
    val amenities: List<String> = emptyList(),
    
    @Convert(converter = StringArrayConverter::class)
    val tags: List<String> = emptyList(),
    
    @Column(name = "transportation_car_time")
    val transportationCarTime: String? = null,
    
    @Column(name = "transportation_bus_time")
    val transportationBusTime: String? = null,
    
    @Convert(converter = StringArrayConverter::class)
    @Column(name = "weather_tags")
    val weatherTags: List<String> = emptyList(),
    
    @Convert(converter = StringArrayConverter::class)
    @Column(name = "noise_tags")
    val noiseTags: List<String> = emptyList(),
    
    val popularity: Int = 0,
    
    @Column(name = "created_at")
    val createdAt: OffsetDateTime = OffsetDateTime.now(),
    
    @OneToMany(mappedBy = "place", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    val bookmarks: List<Bookmark> = emptyList(),
    
    @OneToMany(mappedBy = "place", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    val prompts: List<Prompt> = emptyList(),
    
    @OneToMany(mappedBy = "place", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    val recentViews: List<RecentView> = emptyList()
)